src/Xamarin.Android.Build.Tasks/Utilities/ManagedResourceParser.cs

```diff
@@ -45,15 +46,38 @@ public CodeTypeDeclaration Parse (string resourceDirectory, IEnumerable<string>
 			styleable = CreateClass ("Styleable");
 			style = CreateClass ("Style");
 
-			foreach (var dir in Directory.GetDirectories (resourceDirectory, "*", SearchOption.TopDirectoryOnly)) {
-				foreach (var file in Directory.GetFiles (dir, "*.*", SearchOption.AllDirectories)) {
-					ProcessResourceFile (file);
-				}
+			// This top most R.txt will contain EVERYTHING we need. including library resources since it represents
+			// the final build.
+			var rTxt = Path.Combine(resourceDirectory, "..", "R.txt");
+			if (File.Exists(rTxt))
+			{
+				ProcessRtxtFile(rTxt);
 			}
-			if (additionalResourceDirectories != null) {
-				foreach (var dir in additionalResourceDirectories) {
-					foreach (var file in Directory.GetFiles (dir, "*.*", SearchOption.AllDirectories)) {
-						ProcessResourceFile (file);
+			else
+			{
+				foreach (var dir in Directory.GetDirectories(resourceDirectory, "*", SearchOption.TopDirectoryOnly))
+				{
+					foreach (var file in Directory.GetFiles(dir, "*.*", SearchOption.AllDirectories))
+					{
+						ProcessResourceFile(file);
+					}
+				}
+				if (additionalResourceDirectories != null)
+				{
+					foreach (var dir in additionalResourceDirectories)
+					{
+						rTxt = Path.Combine(dir, "..", "R.txt");
+						if (File.Exists(rTxt))
```

Comment:
This feels duplicative of the above code, in that the `foreach (var dir in additionalResourceDirectories)` block appears to be nearly identical to the previous block. Is there any way to "flatten" these?

```csharp
foreach (var dir in new[]{resourceDirectory}.Concat(additionalResourceDirectories ?? new string[0])) {
    var rTxt = Path.Combine (dir, "..", "R.txt");
    if (File.Exists (rTxt)) {
        ProcessRtxtFile (rTxt);
        continue;
    }
    foreach (var file in Directory.GetFiles (dir, "*.*", SearchOption.AllDirectories)) {
        ProcessResourceFile (file);
    }
}
```

Reply:
One of the reasons  I split this up like this is because if we find a `R.txt` in the `resourceDirectory` that will have been generated by the app. As a result it will contain EVERYTHING. There will be no need to process anything else. 

If its not there then we need to process everything else. 

For the `additionalResourceDirectories` if we find an `R.txt` we can use that. but it will only contain the resources ids for that directory.. not everything. 

Now I do have checks to make sure we don't add duplicates.. but this needs to be quick. And the quickest way to read a file is not too. So the more we can skip over the better.